container-core/src/main/java/com/yahoo/container/handler/LogHandler.java

```diff
@@ -60,26 +61,57 @@ public void render(OutputStream output, ContentChannel networkChannel, Completio
     }
 
 
-    private static class BlockingFlushContentChannelOutputStream extends ContentChannelOutputStream {
-
+    private static class MaxPendingContentChannelOutputStream extends ContentChannelOutputStream {
         private final ContentChannel channel;
+        private final long maxPending;
+        private AtomicLong sent = new AtomicLong(0);
+        private AtomicLong acked = new AtomicLong(0);
 
-        public BlockingFlushContentChannelOutputStream(ContentChannel endpoint) {
+        public MaxPendingContentChannelOutputStream(ContentChannel endpoint, long maxPending) {
             super(endpoint);
             this.channel = endpoint;
+            this.maxPending = maxPending;
+        }
+
+        private long pendingBytes() {
+            return sent.get() - acked.get();
+        }
+
+        private class TrackCompletition implements CompletionHandler {
+            final long written;
+            TrackCompletition(long written) {
+                this.written = written;
+                sent.addAndGet(written);
+            }
+            @Override
+            public void completed() {
+                acked.addAndGet(written);
+            }
+
+            @Override
+            public void failed(Throwable t) {
+                acked.addAndGet(written);
+            }
+        }
+        @Override
+        public void send(ByteBuffer src) throws IOException {
+            try {
+                stallWhilePendingAbove(maxPending);
+            } catch (InterruptedException e) {}
+            send(src, new TrackCompletition(src.remaining()));
+        }
+
+        private void stallWhilePendingAbove(long pending) throws InterruptedException {
+            while (pendingBytes() > pending) {
+                Thread.sleep(1);
```

Comment:
Perhaps a `java.util.concurrent.Condition` would be preferable over `Thread.sleep()`?

Reply:
Hmm, not sure, after all.

Reply:
If you use signalling you increase the cost as that is around 3x more costly than an atomic add operation. We do not need the immediate wakeup as the inflight amount is significantly larger than a 1ms window. In these cases I find it preferable to use sleep.

Reply:
Yup. Thanks for the numbers.